source/s2n/tls_channel_handler.c

```diff
@@ -0,0 +1,738 @@
+    s2n_handler->latest_message_completion_user_data = message->user_data;
+
+    s2n_blocked_status blocked;
+    ssize_t write_code =
+        s2n_send(s2n_handler->connection, message->message_data.buffer, (ssize_t)message->message_data.len, &blocked);
+
+    ssize_t message_len = (ssize_t)message->message_data.len;
+    aws_channel_release_message_to_pool(slot->channel, message);
+
+    if (write_code < message_len) {
+        return aws_raise_error(AWS_IO_TLS_ERROR_WRITE_FAILURE);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_s2n_handler_shutdown(
+    struct aws_channel_handler *handler,
+    struct aws_channel_slot *slot,
+    enum aws_channel_direction dir,
+    int error_code,
+    bool abort_immediately) {
+    struct s2n_handler *s2n_handler = (struct s2n_handler *)handler->impl;
+
+    if (dir == AWS_CHANNEL_DIR_WRITE && !error_code) {
+        s2n_blocked_status blocked;
+        /* make a best effort, but the channel is going away after this run, so.... you only get one shot anyways */
+        s2n_shutdown(s2n_handler->connection, &blocked);
+    } else {
+        while (!aws_linked_list_empty(&s2n_handler->input_queue)) {
+            struct aws_linked_list_node *node = aws_linked_list_pop_front(&s2n_handler->input_queue);
+            struct aws_io_message *message = AWS_CONTAINER_OF(node, struct aws_io_message, queueing_handle);
+            aws_channel_release_message_to_pool(s2n_handler->slot->channel, message);
+        }
+    }
+
+    return aws_channel_slot_on_handler_shutdown_complete(slot, dir, error_code, abort_immediately);
+}
+
+static void s_run_read(struct aws_task *task, void *arg, aws_task_status status) {
+    task->fn = NULL;
+    task->arg = NULL;
+
+    if (status == AWS_TASK_STATUS_RUN_READY) {
+        struct aws_channel_handler *handler = (struct aws_channel_handler *)arg;
+        struct s2n_handler *s2n_handler = (struct s2n_handler *)handler->impl;
+        s_s2n_handler_process_read_message(handler, s2n_handler->slot, NULL);
+    }
+}
+
+static int s_s2n_handler_increment_read_window(
+    struct aws_channel_handler *handler,
+    struct aws_channel_slot *slot,
+    size_t size) {
+    (void)size;
+    struct s2n_handler *s2n_handler = (struct s2n_handler *)handler->impl;
+
+    size_t downstream_size = aws_channel_slot_downstream_read_window(slot);
+    size_t current_window_size = slot->window_size;
+
+    if (downstream_size <= current_window_size) {
+        size_t likely_records_count = (downstream_size - current_window_size) % MAX_RECORD_SIZE;
+        size_t offset_size = likely_records_count * (EST_TLS_RECORD_OVERHEAD);
+        size_t window_update_size = (downstream_size - current_window_size) + offset_size;
+        aws_channel_slot_increment_read_window(slot, window_update_size);
+    }
+
+    if (s2n_handler->negotiation_finished) {
+        /* we have messages in a queue and they need to be run after the socket has popped (even if it didn't have data
```

Comment:
Same as the other place where this comment is copied from, this comment could be more clear